switools/signaturelib.py

```diff
@@ -12,4 +14,45 @@ def getSigFileName( swiFile ):
       return SWIX_SIG_FILE_NAME
    return SWI_SIG_FILE_NAME
 
+def runCmd( cmd, workDir ):
+   try:
+      subprocess.check_call( cmd.split(" "), cwd=workDir )
+   except subprocess.CalledProcessError:
+      return False
+   return True
 
+def getOptimizations( swi, workDir ):
+   # unzip spits warnings when extracting non-existant files, so check first :-(
+   cmd = "unzip -Z1 %s" % swi
+   if not "swimSqshMap" in str( subprocess.check_output( cmd.split(" ") ) ):
+      return None # legacy image
+   cmd = "unzip -qq -o %s swimSqshMap" % os.path.abspath( swi )
+   try:
+      subprocess.check_call( cmd.split(" "), cwd=workDir )
+   except subprocess.CalledProcessError:
+      return None # legacy image
+
+   optims = []
+   with open( "%s/swimSqshMap" % workDir ) as f:
+      for line in f:
+         optim, _ = line.split( "=", 1 )
+         optims.append( optim )
+   return optims
+
+def extractSwadapt( swi, workDir ):
+   cmd = "unzip -o -qq %s swadapt" % os.path.abspath( swi )
```

Comment:
Please use `zipfile`.

Reply:
Forget that, `zipfile` doesn't restore the execute permissions, and although you can workaround that with `os.chmod` it's a bit ugly (although whether it's more ugly than running `unzip` is debatable 😛 )